### PR TITLE
Convert mastery page to react, allow pagination through all players

### PR DESCRIPTION
**Pull request by Jamiras**
image
image

This addresses the "Allow Most Points Earned widget to display 'more...'" bullet point in the automated achievement of the week plan.

### CODE REVIEW

File: app/Platform/Controllers/GameTopAchieversController.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+        $topAchieversService->initialize($game);
+        $paginatedUsers = $topAchieversService->highestPointEarnersQuery()->paginate($perPage);
+
+        $items = [];
+        $rank = 0;
+        $rankScore = -1;
+        $firstRank = 0;
+        $nextRank = 0;
+        foreach ($paginatedUsers->getCollection() as $playerGame) {
```
Comment (wescopeland) [inline app/Platform/Controllers/GameTopAchieversController.php:52]: nit: My VSCode seems very unhappy with `getCollection()` here (getting a red squiggly). The red squiggly goes away if this is changed to `$paginatedUsers->items()`.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+} from '@/common/components/+vendor/BaseTable';
+import { PlayerBadgeIndicator } from '@/common/components/PlayerBadgeIndicator';
+import { PlayerBadgeLabel } from '@/common/components/PlayerBadgeLabel';
+import { UserAvatar } from '@/common/components/UserAvatar';
+import { usePageProps } from '@/common/hooks/usePageProps';
+import { AwardType } from '@/common/utils/generatedAppConstants';
+import { formatDate } from '@/common/utils/l10n/formatDate';
+import { cn } from '@/utils/cn';
+
+export const TopAchieversListContainerTestId = 'top-achievers-list';
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:20]: nit: Conventionally this should be either camelCased or SCREAMING_SNAKE_CASED.

Thankfully, given this is TS, Rename Symbol works really well.
Screenshot 2024-11-23 at 6 33 18 PM

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    return null;
+  }
+
+  return (
+    <BaseTable
+      containerClassName={cn(
+        'overflow-auto rounded-md border border-neutral-700/80 bg-embed',
+        'light:border-neutral-300 lg:overflow-visible lg:rounded-sm',
+        '',
+        '',
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:37]: nit: Safe to delete these two lines.

`cn()` ultimately just produces a concatenated string. It accepts an infinite number of args.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+          <BaseTableHead>{t('Rank')}</BaseTableHead>
+          <BaseTableHead>{t('User')}</BaseTableHead>
+          <BaseTableHead>{t('Progress')}</BaseTableHead>
+        </BaseTableRow>
+      </BaseTableHeader>
+
+      <BaseTableBody>
+        {paginatedUsers.items.map((achiever) => (
+          <BaseTableRow key={achiever.user.displayName}>
+            <BaseTableCell>{achiever.rank}</BaseTableCell>
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:51]: This should pass through the number formatter, eg:

```tsx
const { formatNumber } = useFormatNumber();

<BaseTableCell>{formatNumber(achiever.rank)}</BaseTableCell>
```

otherwise number separators will not work correctly for any locale, ie:

Screenshot 2024-11-23 at 6 37 00 PM

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+            <BaseTableCell>
+              <UserAvatar {...achiever.user} size={32} />
+            </BaseTableCell>
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:55]: There's a curious CSS bug here:

Screenshot 2024-11-23 at 6 38 03 PM

This can be resolved by:
```tsx
<BaseTableCell>
  <div className="max-w-fit">
    <UserAvatar {...achiever.user} size={32} />
  </div>
</BaseTableCell>
```

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+            <BaseTableCell>
+              {achiever.badge ? (
+                <div>
+                  {achiever.badge.awardType == AwardType.Mastery && (
+                    <span>{formatDate(achiever.badge.awardDate, 'lll')}</span>
+                  )}
+
+                  <div className={cn('flex items-center gap-1')}>
+                    {achiever.badge.awardType == AwardType.GameBeaten && (
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:65]: These should both use `===` for improved type safety. This should ideally be a linting rule - I will investigate that.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+            </BaseTableCell>
+
+            <BaseTableCell>
+              {achiever.badge ? (
+                <div>
+                  {achiever.badge.awardType == AwardType.Mastery && (
+                    <span>{formatDate(achiever.badge.awardDate, 'lll')}</span>
+                  )}
+
+                  <div className={cn('flex items-center gap-1')}>
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:64]: nit: This call to `cn()` can be removed. I like to use `cn()` when there's a lot of classes and I want to show them on multiple lines (for readability), or if there's conditional logic driving the classes.

```tsx
<div className="flex items-center gap-1">
```

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                    {achiever.badge.awardType == AwardType.GameBeaten && (
+                      <span>
+                        {achiever.score}
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:67]: This may also need to pass through the number formatter, as with some of the larger sets a player may have 1,000+ points but still no mastery.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                        <span className="text-muted">{' - '}</span>
+                      </span>
+                    )}
+
+                    <PlayerBadgeIndicator playerBadge={achiever.badge} className="mt-px" />
+                    <PlayerBadgeLabel playerBadge={achiever.badge} variant="muted-group" />
+                  </div>
+                </div>
+              ) : (
+                <span>{achiever.score}</span>
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx:77]: Similar feedback here as the other `score` values - this should probably pass through the number formatter.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import { TopAchieversList, TopAchieversListContainerTestId } from './TopAchieversList';
+
+export const createAchiever = createFactory<App.Platform.Data.GameTopAchiever>((faker) => {
+  return {
+    rank: 1,
+    user: createUser(),
+    score: faker.number.int({ min: 1, max: 100 }),
+    badge: null,
+  };
+});
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx:15]: I think this should probably live in a _test/factories/createGameTopAchiever.ts_ file. When the game page is eventually converted to React, I suspect it'll use the same `GameTopAchieverData` DTO for the component that leads to this page.

In other words, you should be able to also import `createGameTopAchiever` from `@/test/factories`. You'll need to add it to _test/factories/index.ts_ and it should just work.

It's ok for the new _createGameTopAchiever.ts_ file to import `createUser`.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  it('given there are no achievers, renders nothing', () => {
+    // ARRANGE
+    render<App.Platform.Data.GameTopAchieversPageProps>(<TopAchieversList />, {
+      pageProps: {
+        paginatedUsers: createPaginatedData([]),
+      },
+    });
+
+    // ASSERT
+    expect(screen.queryByTestId(TopAchieversListContainerTestId)).not.toBeInTheDocument();
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx:42]: I believe this test will always pass.

I don't think the test ID is currently being used on a component. The test id has to be added to a tag, like:

```tsx
<BaseTable
  containerClassName="some stuff"
  data-testid={TopAchieversListContainerTestId}
>
```

This is all an edge case, but I'm wondering if in the event there are no players to display, rather than showing a table, we should display an `<EmptyState>` component (see: _RecentForumPosts.tsx_).

Then the test could be something like:
```tsx
// ASSERT
expect(screen.queryByRole('table')).not.toBeInTheDocument();
expect(screen.getByRole(/no top achievers/i)).toBeVisible();
```

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.test.tsx [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+import { TopAchieversMainRoot } from './TopAchieversMainRoot';
+
+export const createAchiever = createFactory<App.Platform.Data.GameTopAchiever>((faker) => {
+  return {
+    rank: 1,
+    user: createUser(),
+    score: faker.number.int({ min: 1, max: 100 }),
+    badge: null,
+  };
+});
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.test.tsx:17]: See other feedback related to creating a common factory in `@/test/factories`.

**Review by wescopeland — COMMENTED** [review]
Good work, it made me really happy to see this PR come through. I've verified functionality & I've also verified the front-end code has 100% coverage.

I have a few pieces of feedback. If you get stuck on anything please let me know and I am happy to help.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.tsx [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  };
+
+  return (
+    <div>
+      <GameBreadcrumbs
+        game={game}
+        system={game.system}
+        t_currentPageLabel={t('Game Top Achievers')}
+      />
+      <GameHeading game={game}>{t('Game Top Achievers')}</GameHeading>
```
Comment (wescopeland) [inline resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.tsx:33]: minor: I'm wondering if "Top Achievers" as a label may make more sense than "Game Top Achievers".

Given these labels are part of either the game breadcrumbs or the game heading, "Game" in the label could perhaps be redundant.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 2 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: Based on https://github.com/RetroAchievements/RAWeb/blob/31b93029cf412c817d123e308948a8e4b7925179/resources/js/features/games/components/HashesMainRoot/HashesList/HashesList.tsx#L8

Removed.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx (hunk 11 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx]: I changed it to look for "Rank" not being present and removed the TestID.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 3 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: done

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 4 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: done

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 5 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: Not sure what the bug is, but I made the suggested change.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 7 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: It's certainly copied from somewhere, I didn't know what `cn()` did, so I wouldn't have tried to use it. I'm also not great with css, so I probably would have left the class of entirely.

Updated.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 6 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: I have a hard time remembering to do this as it's different from every other language I use. Changed.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 8 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: done.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 9 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: done

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.test.tsx (hunk 12 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.test.tsx]: done.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.tsx (hunk 13 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversMainRoot.tsx]: Agreed. Changed.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx (hunk 10 above)
Comment (Jamiras) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.test.tsx]: done.

File: app/Platform/Controllers/GameTopAchieversController.php (hunk 1 above)
Comment (wescopeland) [reply app/Platform/Controllers/GameTopAchieversController.php]: Looks like the nuance comes down to the underlying types between the two (the one in `BuildGameListAction` isn't complaining for me).

In GameTopAchieversController, `$paginatedUsers` has a type of:
```
\Illuminate\Contracts\Pagination\LengthAwarePaginator $paginatedUsers
```

In BuildGameListAction, `$entries` has a type of:
```
\Illuminate\Pagination\LengthAwarePaginator<\App\Models\Game> $entries
```

One is Contracts\Pagination, whereas the other is just Pagination. Not sure why.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 5 above)
Comment (wescopeland) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: The bug was no matter how far across the table cell I was hovering, the user tooltip card was still appearing.

File: resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx (hunk 6 above)
Comment (wescopeland) [reply resources/js/features/games/components/TopAchieversMainRoot/TopAchieversList/TopAchieversList.tsx]: Yep, and personally I feel you shouldn't have to remember to do it, the linter should just take care of it. This does occur in other projects I work on, so I'll figure out why it isn't working here in a subsequent PR.

Ideally, assuming you have format+fix on save enabled in VSCode, this should just fix itself on saving a file.

**Review by wescopeland — APPROVED** [review]
